src/io/sb3/toSb3.ts

```diff
@@ -1,163 +1,1114 @@
+    // arrays or referenced by block ID is described in serializeInputShadow.
+    // As far as the input mapping is concerned, all that matters is that the
+    // two formats may be interchanged with one another.
+    //
+    // Also note that there are a couple blocks (specifically the variable and
+    // list-contents getters) which are serialized altogether in the compressed
+    // much the same as a compressed shadow, irregardless of the type of the
+    // input they've been placed inside. This is because they're so common in
+    // a project, and do not have any inputs of their own - only a field to
+    // identify which variable or list the block corresponds to. The input
+    // mapping for the block "set x to (spawn x)" would look something like
+    // this:
+    //
+    //   inputs: {
+    //     X: [INPUT_DIFF_BLOCK_SHADOW, [12, "spawn x", "someId"], [4, 0]]
+    //   }
+    //
+    // ...where someId is the ID of the variable, and [4, 0] is the obscured
+    // shadow block, as usual.
+
+    const { block, getBroadcastId, getCustomBlockData, initialValues, inputEntries, stage, target } = options;
+
+    const blockData = newBlockData();
+
+    const result: {
+      inputs: sb3.Block["inputs"];
+      blockData: BlockData;
+    } = {
+      inputs: {},
+      blockData
+    };
+
+    for (const [key, entry] of Object.entries(inputEntries)) {
+      const input = inputs[key];
+      if (entry === sb3.BooleanOrSubstackInputStatus) {
+        let firstBlock: Block;
+        let siblingBlocks: Block[];
+
+        if (input && input.type === "blocks") {
+          firstBlock = input.value[0];
+          siblingBlocks = input.value;
+        } else if (input && input.type === "block") {
+          firstBlock = input.value;
+        }
+
+        if (firstBlock) {
+          const { blockData: inputBlockData, blockId } = serializeBlock(firstBlock, {
+            getBroadcastId,
+            getCustomBlockData,
+            parent: block,
+            siblingBlocks,
+            stage,
+            target
+          });
+          applyBlockData(blockData, inputBlockData);
+
+          if (blockId) {
+            result.inputs[key] = [BIS.INPUT_BLOCK_NO_SHADOW, blockId];
+          }
+        }
+      } else {
+        let valueForShadow;
+        if (input.type === "block") {
+          valueForShadow = initialValues[key];
+          // Special-case some input opcodes for more realistic initial values.
+          if (entry === OpCode.looks_costume) {
+            if (target.costumes[0]) {
+              valueForShadow = target.costumes[0].name;
+            }
+          } else if (entry === OpCode.sound_sounds_menu) {
+            if (target.sounds[0]) {
+              valueForShadow = target.sounds[0].name;
+            }
+          } else if (entry === OpCode.event_broadcast_menu) {
+            valueForShadow = getBroadcastId.initialBroadcastName;
+          }
```

Comment:
```suggestion
          // Special-case some input opcodes for more realistic initial values.
          switch (entry) {
            case OpCode.looks_costume:
              valueForShadow = target.costumes[0].name;
              break;
            case OpCode.sound_sounds_menu:
              valueForShadow = target.sounds[0].name;
              break;
            case OpCode.event_broadcast_menu:
              valueForShadow = getBroadcastId.initialBroadcastName;
              break;
            default:
              valueForShadow = initialValues[key];
              break;
          }
```

Reply:
Also, what happens to `valueForShadow` if the target has no sounds (or no `initialBroadcastName`)? Will it throw an error since it's trying to access `name`?